src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -348,9 +348,16 @@ static inline int MPIDI_OFI_do_am_isend(int rank,
     mpi_errno = MPIDI_OFI_am_init_request(am_hdr, am_hdr_sz, sreq);
     MPIR_ERR_CHECK(mpi_errno);
 
-    if (!dt_contig) {
-        send_buf = (char *) MPL_malloc(data_sz, MPL_MEM_BUFFER);
-
+    MPL_pointer_attr_t attr;
+    MPL_gpu_query_pointer_attr(buf, &attr);
+
+    if (!dt_contig || (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM)) {
+        /* FIXME: currently we always do packing, also for high density types. However,
+         * we should not do packing in those case. For contiguous types we should not
+         * allocate the entire buffer and do the transfer at once. */
+        /* TODO: (1) Skip packing for high-density datatypes;
+         *       (2) Pipeline allocation for contiguous GPU buffers; */
+        MPL_gpu_malloc_host(&send_buf, data_sz);
```

Comment:
@shawnccx @pavanbalaji pointed out we don't need to allocate a `host_buf` as we already have a `pack_buffer` field in AM request. We can simply always allocate the `pack_buffer` with `MPL_gpu_malloc_host`, adding the packing condition also for contig GPU buffers if the network layer does not supports them. Can you check if the same can be done for OFI direct netmod. If that is the case we just need to handle the recv case.

Reply:
Actually, in OFI we can do similar thing also for recv: https://github.com/pmodels/mpich/blob/2a04eadd2cfb233e3382d76af0c242ed07181cec/src/mpid/ch4/netmod/ofi/ofi_recv.h#L176

For UCX things are a little different.

Reply:
It seems doable to me, I can add this later.

Reply:
If you can change ofi code to only use ofi pack/unpack buffers we may not need any of the `mpidig_gpu_utils` at all.